File: algorithmSolution/leetCodeC++/007_Reverse Integer.cpp
/*   DwyaneTalk@gmail.com
 *   by lutao
 *   leetCodeC++	: Reverse Integer
 *					: https://leetcode.com/problems/reverse-integer/
*/
#include <iostream>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
	int reverse(int x) {
		int sign = (x >= 0 ? 1 : -1), ans = 0;
		if(x == INT_MIN || x == 0)	return 0;
		x = (x >= 0 ? x : -x);
		while(!(x % 10))	x = x / 10;
		while(x) {
			if(sign) {
				if(ans > (INT_MAX - x % 10) / 10)	return 0;
			}else {
				if(ans > -(INT_MIN + x % 10) / 10)	return 0;
			}
			ans = ans * 10 + (x % 10);
			x = x / 10;
		}
		ans *= sign;
		return ans;
	}

	int reverse_new(int x) {
		register int ans = 0;
		while(x) {
			int newAns = ans * 10 + x % 10;
			if(newAns / 10 != ans)	return 0;
			ans = newAns;
			x /= 10;
		}
		return ans;
	}
};

File: algorithmSolution/leetCodePython/007_Reverse Integer.py
'''
DwyaneTalk@gmail.com
https://leetcode.com/problems/two-sum/
'''
class Solution(object):
    def reverse(self, x):
		flags = 1
		if x < 0:
			flags = -1
			x = -x
		ans = 0
		while x :
			ans = ans * 10 + x % 10
			x = x / 10
		if (flags > 0 and ans > (1 << 32) - 1) or (flags < 0 and ans > (1 << 32)):
			return 0
		return  ans * flags